crates/ruma-events/src/state_key.rs

```diff
@@ -38,3 +39,208 @@ impl Serialize for EmptyStateKey {
+    ///   true`
+    /// - "{UserId}_{DeviceId}" example: "@test:user.org_DEVICE". `device_id`: Some`, `underscore:
+    ///   false`
+    /// - "{UserId}" example example: "@test:user.org". `device_id`: None`, underscore is ignored:
+    ///   `underscore: false|true`
+    ///
+    /// Dependent on the parameters the correct CallMemberStateKey will be constructed.
+    pub fn new(user_id: OwnedUserId, device_id: Option<OwnedDeviceId>, underscore: bool) -> Self {
+        CallMemberStateKeyEnum::new(user_id, device_id, underscore).into()
+    }
+
+    /// Returns the user id in this state key.
+    /// (This is a cheap operations. The id is already type checked on initialization. And does only
+    /// returns a reference to an existing OwnedUserId.)
+    pub fn user_id(&self) -> &UserId {
+        match &self.key {
+            CallMemberStateKeyEnum::UnderscoreUserDevice(u, _) => u,
+            CallMemberStateKeyEnum::UserDevice(u, _) => u,
+            CallMemberStateKeyEnum::User(u) => u,
+        }
+        .as_ref()
+    }
+
+    /// Returns the device id in this state key (if available)
+    /// (This is a cheap operations. The id is already type checked on initialization. And does only
+    /// returns a reference to an existing OwnedDeviceId.)
+    pub fn device_id(&self) -> Option<&DeviceId> {
+        match &self.key {
+            CallMemberStateKeyEnum::UnderscoreUserDevice(_, d) => Some(d),
+            CallMemberStateKeyEnum::UserDevice(_, d) => Some(d),
+            CallMemberStateKeyEnum::User(_) => None,
+        }
+        .map(|d| d.as_ref())
+    }
+}
+
+impl AsRef<str> for CallMemberStateKey {
+    fn as_ref(&self) -> &str {
+        &self.raw
+    }
+}
+
+impl From<CallMemberStateKeyEnum> for CallMemberStateKey {
+    fn from(value: CallMemberStateKeyEnum) -> Self {
+        let raw: Box<str> = match &value {
+            CallMemberStateKeyEnum::UnderscoreUserDevice(u, d) => format!("_{}_{}", u, d).into(),
+            CallMemberStateKeyEnum::UserDevice(u, d) => format!("{}_{}", u, d).into(),
+            CallMemberStateKeyEnum::User(u) => u.to_string().into(),
+        };
+        Self { key: value, raw }
+    }
+}
+
+impl FromStr for CallMemberStateKey {
+    type Err = MemberParseError;
+
+    fn from_str(state_key: &str) -> Result<Self, Self::Err> {
+        // Intentionally do not use CallMemberStateKeyEnum.into since this would reconstruct the
+        // state key string.
+        Ok(Self { key: CallMemberStateKeyEnum::from_str(state_key)?, raw: state_key.into() })
+    }
+}
+
+impl<'de> Deserialize<'de> for CallMemberStateKey {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = ruma_common::serde::deserialize_cow_str(deserializer)?;
+        let string = String::from(s.as_ref());
+        Self::from_str(&string).map_err(|err| de::Error::invalid_value(Unexpected::Str(&s), &err))
```

Comment:
The first line makes an unnecessary allocation. This should work:

```suggestion
        Self::from_str(&s).map_err(|err| de::Error::invalid_value(Unexpected::Str(&s), &err))
```